akka-projection-testkit/src/main/scala/akka/projection/testkit/ProjectionTestKit.scala

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.projection.testkit
+
+import akka.Done
+import akka.actor.testkit.typed.TestKitSettings
+import akka.actor.testkit.typed.scaladsl.{ ActorTestKit, _ }
+import akka.annotation.ApiMayChange
+import akka.projection.Projection
+
+import scala.concurrent.duration._
+import scala.concurrent.{ Future, Promise }
+@ApiMayChange
+object ProjectionTestKit {
+  def apply(testKit: ActorTestKit): ProjectionTestKit =
+    new ProjectionTestKit(testKit)
+}
+
+@ApiMayChange
+final class ProjectionTestKit private[akka] (testKit: ActorTestKit) {
+
+  private implicit val system = testKit.system
+  private implicit val settings: TestKitSettings = TestKitSettings(system)
+
+  def run(proj: Projection)(testFunc: => Unit): Unit =
+    runInternal(proj, testFunc, settings.SingleExpectDefaultTimeout.dilated, 100.millis)
+
+  def run(proj: Projection, max: FiniteDuration)(testFunc: => Unit): Unit = {
+    runInternal(proj, testFunc, max, 100.millis)
+  }
+
+  def run(proj: Projection, max: FiniteDuration, interval: FiniteDuration = 100.millis)(testFunc: => Unit): Unit = {
+    runInternal(proj, testFunc, max, 100.millis)
+  }
+
+  private def runInternal(
```

Comment:
there are three variants of `run` that follows the `Probe.awaitAssert` signature. 

As such, users don't need to wrap their test function with a `ScalaTest.eventually` or similar.

Reply:
Question is if the returned `Future` is of much value for the tests? Could cause confusion since the run is otherwise blocking (which is convenient for tests). Maybe it should be `run(...): Unit` and we Await the stop Future?

Reply:
I see what you mean. Make sense.

Reply:
Since the run is otherwise blocking (in a good way) I find it strange to mix with a Future return value. Now we throw such exception, which I think is better because then it's not silently ignored. If testing for that it should throw one can use `intercept` in Scalatest.

Reply:
I agree that returning a `Future` doesn't make sense, but if an exception occurs within the projection's user event handler `Projection.processElement` won't it will be swallowed in the stream completion since the mat value is not accessible to the caller of the testkit?

Reply:
I think the idea was that `stop()` would return the `Future[Done]` mat value (containing the exception). Since we Await stop that will be thrown.

Reply:
I tested this out by introducing an exception into the event handler, but a `org.scalatest.exceptions.TestFailedException` exception raised by `probe.awaitAssert` is thrown before the projection is stopped, because the assertion predicate was never satisfied before the timeout.

Reply:
ok, that's a bug then. I think it's because it's using `Await.ready(proj.stop(), max)`. Should be `Await.result(proj.stop(), max)` to throw

Reply:
I think there were some misunderstanding here and maybe that whole approach will always create confusion and therefore we should re-think it. 

If the stream fails, the `assertFunc` will probably fail and the user will get a `TestFailedException`. The stream failure becomes irrelevant and is eclipsed by the user function. 

The goal is not to test the stream, but the user's event-handler. 

Since projections are background processes, a stream failure will cause the stream to stop and re-start (once we add a back-off mechanism). 

What we could do instead is to catch `TestFailedException` and only throw it if the stream completes/stops without a failure. Meaning, the test failed because the `assertFunc` failed. 

If the streams ends with a failure, we swallow the `TestFailedException` because it becomes less relevant then the stream failure. Instead, we throw the stream failure.

Reply:
> I think there were some misunderstanding here and maybe that whole approach will always create confusion and therefore we should re-think it.

After thinking about this a little more I understand where I got confused.  If the primary use case for using this is to mimic `eventually` behaviour because you want to _eventually_ see some state from the projection then exceptions of any kind are irrelevant (unless we're eventually waiting for an exception).

Am I correct in assuming that the stream running a projections should never complete, unless there's an exception or they're shutdown externally?

> What we could do instead is to catch `TestFailedException` and only throw it if the stream completes/stops without a failure. Meaning, the test failed because the `assertFunc` failed.
> 
> If the streams ends with a failure, we swallow the `TestFailedException` because it becomes less relevant then the stream failure. Instead, we throw the stream failure.

That makes sense to me.

Reply:
Updated the code by doing exactly what @patriknw suggested, ie: `Await.result`. 

Added two more tests. One with exception inside the user function and one with an exception in the stream.

Reply:
> code by doing exactly what @patriknw suggested

what I mean by that is that I had something else in mind, but it turns out that using `Await.result` solves it in a much simpler way.

Reply:
> Am I correct in assuming that the stream running a projections should never complete, unless there's an exception or they're shutdown externally?

Yes, that correct. 

My last changes preserve the original idea. However, if for some reason there is a failure in the stream, the user will now see it. Before we were swallowing it and the user was always getting a `TestFailedException`.